acpi/src/lib.rs

```diff
@@ -95,189 +100,173 @@ pub enum AcpiError {
+                    rsdp_mapping.rsdt_address() as usize,
+                    core::mem::size_of::<SdtHeader>(),
+                )
+            };
+            rsdt_mapping.validate(Signature::RSDT)?;
+
+            Ok(Self { mapping: rsdt_mapping, revision, handler })
         } else {
             /*
              * We're running on ACPI Version 2.0+. We should use the 64-bit XSDT address, truncated
              * to 32 bits on x86.
              */
-            let xsdt_address = rsdp_mapping.xsdt_address();
-            unsafe { Self::from_rsdt(handler, revision, xsdt_address as usize) }
+
+            // ### Safety: Addresses from a validated `RSDP` are also guaranteed to be valid.
+            let xsdt_mapping: PhysicalMapping<H, SdtHeader> = unsafe {
+                handler.map_physical_region::<SdtHeader>(
+                    rsdp_mapping.xsdt_address() as usize,
+                    core::mem::size_of::<SdtHeader>(),
+                )
+            };
+            xsdt_mapping.validate(Signature::XSDT)?;
+
+            Ok(Self { mapping: xsdt_mapping, revision, handler })
         }
     }
 
-    /// Create an `AcpiTables` if you have the physical address of the RSDT. This is useful, for example, if your chosen
-    /// bootloader reads the RSDP and passes you the address of the RSDT. You also need to supply the correct ACPI
-    /// revision - if `0`, a RSDT is expected, while a `XSDT` is expected for greater revisions.
-    pub unsafe fn from_rsdt(handler: H, revision: u8, rsdt_address: usize) -> Result<AcpiTables<H>, AcpiError> {
-        let mut result = AcpiTables { revision, sdts: BTreeMap::new(), dsdt: None, ssdts: Vec::new(), handler };
-
-        let header = sdt::peek_at_sdt_header(&result.handler, rsdt_address);
-        let mapping =
-            unsafe { result.handler.map_physical_region::<SdtHeader>(rsdt_address, header.length as usize) };
+    /// Searches through the ACPI table hears and attempts to locate the table with a matching `T::SIGNATURE`.
+    pub fn find_table<T: AcpiTable>(&self) -> AcpiResult<PhysicalMapping<H, T>> {
+        use core::mem::size_of;
 
-        if revision == 0 {
-            /*
-             * ACPI Version 1.0. It's a RSDT!
-             */
-            mapping.validate(sdt::Signature::RSDT)?;
+        if self.revision == 0 {
+            let num_tables = ((self.mapping.length as usize) - size_of::<SdtHeader>()) / size_of::<u32>();
+            // ### Safety: Table pointer is known-good for these offsets and types.
+            let tables_base = unsafe { self.mapping.virtual_start().as_ptr().add(1).cast::<u32>() };
 
-            let num_tables = (mapping.length as usize - mem::size_of::<SdtHeader>()) / mem::size_of::<u32>();
-            let tables_base =
-                ((mapping.virtual_start().as_ptr() as usize) + mem::size_of::<SdtHeader>()) as *const u32;
+            for offset in 0..num_tables {
+                // ### Safety: See above safety message.
```

Comment:
Nit: I'd rather repeat why it's safe than tell the user to refer to a different comment